File: src/storage/migrations.ts
import type Database from 'better-sqlite3';
import {
  EXCHANGE_METADATA_SETTLEMENT_FLOOR_MS,
  EXCHANGE_RESPONSE_ACCEPTANCE_WINDOW_MS,
  LEGACY_EXCHANGE_RETENTION_DAYS,
  RETENTION_DAY_MS,
} from '../domain/exchange-retention.js';
import { classifyStorageError, incompatibleSchema, StorageError } from './errors.js';

type SqliteDatabase = Database.Database;

function saturatingAddMilliseconds(value: number, delta: number, label: string): number {
  if (!Number.isSafeInteger(value) || value <= 0 || !Number.isSafeInteger(delta) || delta < 0) {
    throw new Error(`${label} is outside the supported range.`);
  }
  return value > Number.MAX_SAFE_INTEGER - delta ? Number.MAX_SAFE_INTEGER : value + delta;
}

export interface MigrationDefinition {
  readonly version: number;
  readonly name: string;
  readonly up: (database: SqliteDatabase) => void;
}

/** Domain tables are appended by their owning tickets, in migration order. */
export const CURRENT_MIGRATIONS: readonly MigrationDefinition[] = [
  {
    version: 1,
    name: 'create durable identities and transient tmux bindings',
    up: (database) =>
      database.exec(`
        CREATE TABLE identities (
          id TEXT PRIMARY KEY,
          name TEXT NOT NULL,
          canonical_name TEXT NOT NULL UNIQUE,
          created_at TEXT NOT NULL,
          updated_at TEXT NOT NULL
        );
        CREATE TABLE bindings (
          id TEXT PRIMARY KEY,
          identity_id TEXT NOT NULL REFERENCES identities(id) ON DELETE CASCADE,
          transport TEXT NOT NULL CHECK (transport = 'tmux'),
          pane_id TEXT NOT NULL,
          server_id TEXT NOT NULL,
          socket_path TEXT NOT NULL,
          server_pid INTEGER NOT NULL,
          server_start_time TEXT NOT NULL,
          pane_pid INTEGER NOT NULL,
          bound_at TEXT NOT NULL,
          last_verified_at TEXT NOT NULL,
          UNIQUE(identity_id),
          UNIQUE(transport, server_id, pane_id)
        );
        CREATE INDEX bindings_endpoint ON bindings(transport, server_id, pane_id);
      `),
  },
  {
    version: 2,
    name: 'create optional identity role profiles',
    up: (database) =>
      database.exec(`
        CREATE TABLE role_profiles (
          identity_id TEXT NOT NULL PRIMARY KEY REFERENCES identities(id) ON DELETE CASCADE,
          content TEXT NOT NULL,
          updated_at TEXT NOT NULL
        );
      `),
  },
  {
    version: 3,
    name: 'create durable identity preambles',
    up: (database) =>
      database.exec(`
        CREATE TABLE identity_preambles (
          identity_id TEXT NOT NULL PRIMARY KEY REFERENCES identities(id) ON DELETE CASCADE,
          content TEXT NOT NULL,
          updated_at TEXT NOT NULL
        );
      `),
  },
  {
    version: 4,
    name: 'create request attempts and preamble cadence counters',
    up: (database) =>
      database.exec(`
        CREATE TABLE preamble_counters (
          identity_id TEXT PRIMARY KEY REFERENCES identities(id) ON DELETE CASCADE,
          reserved_count INTEGER NOT NULL CHECK (reserved_count >= 0),
          updated_at_ms INTEGER NOT NULL
        );
        CREATE TABLE request_attempts (
          attempt_id TEXT PRIMARY KEY,
          request_id TEXT NOT NULL UNIQUE,
          nonce TEXT,
          identity_id TEXT REFERENCES identities(id) ON DELETE SET NULL,
          server_id TEXT NOT NULL,
          socket_path TEXT NOT NULL,
          server_pid INTEGER NOT NULL CHECK (server_pid > 0),
          server_start_time TEXT NOT NULL,
          pane_id TEXT NOT NULL,
          pane_pid INTEGER NOT NULL CHECK (pane_pid > 0),
          wait_active INTEGER NOT NULL CHECK (wait_active IN (0, 1)),
          status TEXT NOT NULL CHECK (
            status IN ('prepared', 'sending', 'sent', 'uncertain', 'definitely_failed')
          ),
          preamble_every INTEGER CHECK (preamble_every IS NULL OR preamble_every > 0),
          inject_preamble INTEGER NOT NULL CHECK (inject_preamble IN (0, 1)),
          cadence_reserved INTEGER NOT NULL CHECK (cadence_reserved IN (0, 1)),
          prepared_at_ms INTEGER NOT NULL,
          sending_at_ms INTEGER,
          settled_at_ms INTEGER,
          wait_released_at_ms INTEGER,
          expires_at_ms INTEGER NOT NULL
        );
        CREATE INDEX request_attempts_endpoint_active
          ON request_attempts (
            server_id, socket_path, server_pid, server_start_time, pane_id, pane_pid,
            wait_active, status, prepared_at_ms
          );
        CREATE INDEX request_attempts_retention
          ON request_attempts (wait_active, status, settled_at_ms);
        CREATE INDEX request_attempts_expiry
          ON request_attempts (expires_at_ms, status);
      `),
  },
  {
    version: 5,
    name: 'create immutable request responses',
    up: (database) =>
      database.exec(`
        ALTER TABLE request_attempts
          ADD COLUMN response_submitted_at_ms INTEGER;
        CREATE TABLE request_responses (
          request_id TEXT PRIMARY KEY,
          attempt_id TEXT NOT NULL,
          server_id TEXT NOT NULL,
          socket_path TEXT NOT NULL,
          server_pid INTEGER NOT NULL CHECK (server_pid > 0),
          server_start_time TEXT NOT NULL,
          pane_id TEXT NOT NULL,
          pane_pid INTEGER NOT NULL CHECK (pane_pid > 0),
          body TEXT NOT NULL,
          body_bytes INTEGER NOT NULL CHECK (body_bytes >= 0),
          submitted_at_ms INTEGER NOT NULL CHECK (submitted_at_ms > 0)
        );
        CREATE INDEX request_responses_retention
          ON request_responses (submitted_at_ms);
      `),
  },
  {
    version: 6,
    name: 'freeze exchange retention and response expiry horizons',
    up: (database) => {
      database.exec(`
        ALTER TABLE request_attempts
          ADD COLUMN retention_days INTEGER NOT NULL DEFAULT ${LEGACY_EXCHANGE_RETENTION_DAYS};
        ALTER TABLE request_attempts
          ADD COLUMN retention_expires_at_ms INTEGER NOT NULL DEFAULT 0;
        ALTER TABLE request_responses
          ADD COLUMN response_expires_at_ms INTEGER NOT NULL DEFAULT 0;
      `);

      const selectAttempts = database.prepare(
        `SELECT attempt_id, prepared_at_ms, expires_at_ms, settled_at_ms
         FROM request_attempts
         WHERE attempt_id > ?
         ORDER BY attempt_id
         LIMIT 100`
      );
      const selectFirstAttempts = database.prepare(
        `SELECT attempt_id, prepared_at_ms, expires_at_ms, settled_at_ms
         FROM request_attempts ORDER BY attempt_id LIMIT 100`
      );
      const updateAttempt = database.prepare(
        'UPDATE request_attempts SET retention_expires_at_ms = ? WHERE attempt_id = ?'
      );
      let previousAttemptId: string | undefined;
      while (true) {
        const attempts = (
          previousAttemptId === undefined
            ? selectFirstAttempts.all()
            : selectAttempts.all(previousAttemptId)
        ) as Array<{
          attempt_id: string;
          prepared_at_ms: number;
          expires_at_ms: number;
          settled_at_ms: number | null;
        }>;
        if (attempts.length === 0) break;
        for (const attempt of attempts) {
          const frozenRetention = saturatingAddMilliseconds(
            attempt.prepared_at_ms,
            LEGACY_EXCHANGE_RETENTION_DAYS * RETENTION_DAY_MS,
            `Attempt '${attempt.attempt_id}' retention expiry`
          );
          const responseAcceptance = saturatingAddMilliseconds(
            attempt.prepared_at_ms,
            EXCHANGE_RESPONSE_ACCEPTANCE_WINDOW_MS,
            `Attempt '${attempt.attempt_id}' response acceptance deadline`
          );
          const settlementFloor = saturatingAddMilliseconds(
            attempt.expires_at_ms,
            EXCHANGE_METADATA_SETTLEMENT_FLOOR_MS,
            `Attempt '${attempt.attempt_id}' settlement retention floor`
          );
          const settledFloor =
            attempt.settled_at_ms === null
              ? 0
              : saturatingAddMilliseconds(
                  attempt.settled_at_ms,
                  EXCHANGE_METADATA_SETTLEMENT_FLOOR_MS,
                  `Attempt '${attempt.attempt_id}' settled retention floor`
                );
          updateAttempt.run(
            Math.max(frozenRetention, responseAcceptance, settlementFloor, settledFloor),
            attempt.attempt_id
          );
          previousAttemptId = attempt.attempt_id;
        }
      }

      const selectResponses = database.prepare(
        `SELECT request_id, submitted_at_ms
         FROM request_responses
         WHERE request_id > ?
         ORDER BY request_id
         LIMIT 100`
      );
      const selectFirstResponses = database.prepare(
        `SELECT request_id, submitted_at_ms
         FROM request_responses ORDER BY request_id LIMIT 100`
      );
      const updateResponse = database.prepare(
        'UPDATE request_responses SET response_expires_at_ms = ? WHERE request_id = ?'
      );
      let previousRequestId: string | undefined;
      while (true) {
        const responses = (
          previousRequestId === undefined
            ? selectFirstResponses.all()
            : selectResponses.all(previousRequestId)
        ) as Array<{
          request_id: string;
          submitted_at_ms: number;
        }>;
        if (responses.length === 0) break;
        for (const response of responses) {
          updateResponse.run(
            saturatingAddMilliseconds(
              response.submitted_at_ms,
              LEGACY_EXCHANGE_RETENTION_DAYS * RETENTION_DAY_MS,
              `Response '${response.request_id}' retention expiry`
            ),
            response.request_id
          );
          previousRequestId = response.request_id;
        }
      }
      database.exec(`
        UPDATE request_attempts
        SET retention_expires_at_ms = MAX(
          retention_expires_at_ms,
          COALESCE(
            (SELECT response_expires_at_ms
             FROM request_responses
             WHERE request_responses.request_id = request_attempts.request_id
               AND request_responses.attempt_id = request_attempts.attempt_id),
            retention_expires_at_ms
          )
        )
      `);

      database.exec(`
        CREATE INDEX request_attempts_retention_horizon
          ON request_attempts (retention_expires_at_ms, attempt_id);
        CREATE INDEX request_attempts_cleanup_expiry
          ON request_attempts (expires_at_ms, attempt_id)
          WHERE wait_active = 1 OR status IN ('prepared', 'sending');
        CREATE INDEX request_responses_expiry
          ON request_responses (response_expires_at_ms, request_id);
        CREATE INDEX request_responses_attempt
          ON request_responses (attempt_id, response_expires_at_ms);
      `);
    },
  },
  {
    version: 7,
    name: 'retain request provenance and bounded original prompts',
    up: (database) =>
      database.exec(`
        ALTER TABLE request_attempts
          ADD COLUMN originator_kind TEXT NOT NULL DEFAULT 'unknown'
            CHECK (originator_kind IN ('unknown', 'explicit', 'verified'));
        ALTER TABLE request_attempts
          ADD COLUMN originator_identity_id TEXT
            CHECK (
              (originator_kind = 'unknown' AND originator_identity_id IS NULL) OR
              (originator_kind IN ('explicit', 'verified') AND originator_identity_id IS NOT NULL)
            );
        ALTER TABLE request_attempts
          ADD COLUMN recipient_identity_id TEXT;
        ALTER TABLE request_attempts
          ADD COLUMN message_text TEXT;
        ALTER TABLE request_attempts
          ADD COLUMN message_bytes INTEGER
            CHECK (message_bytes IS NULL OR message_bytes >= 0);
        ALTER TABLE request_attempts
          ADD COLUMN message_expires_at_ms INTEGER;
        CREATE INDEX request_attempts_prompt_expiry
          ON request_attempts (message_expires_at_ms, attempt_id)
          WHERE message_text IS NOT NULL;
      `),
  },
  {
    version: 8,
    name: 'add identity-scoped exchange attention revisions',
    up: (database) =>
      database.exec(`
        ALTER TABLE request_attempts
          ADD COLUMN attention_revision INTEGER NOT NULL DEFAULT 0
            CHECK (
              typeof(attention_revision) = 'integer' AND
              attention_revision >= 0 AND attention_revision <= 9007199254740991
            );
        ALTER TABLE request_attempts
          ADD COLUMN attention_acknowledged_revision INTEGER NOT NULL DEFAULT 0
            CHECK (
              typeof(attention_acknowledged_revision) = 'integer' AND
              attention_acknowledged_revision >= 0 AND
              attention_acknowledged_revision <= 9007199254740991
            );
        CREATE TABLE request_attention_identities (
          identity_id TEXT PRIMARY KEY,
          latest_revision INTEGER NOT NULL CHECK (
            typeof(latest_revision) = 'integer' AND
            latest_revision >= 0 AND latest_revision <= 9007199254740991
          ),
          acknowledged_through INTEGER NOT NULL CHECK (
            typeof(acknowledged_through) = 'integer' AND
            acknowledged_through <= latest_revision AND
            acknowledged_through >= 0 AND acknowledged_through <= 9007199254740991
          )
        );
        WITH ranked AS (
          SELECT attempt_id,
                 ROW_NUMBER() OVER (
                   PARTITION BY originator_identity_id
                   ORDER BY prepared_at_ms, request_id
                 ) AS revision
          FROM request_attempts
          WHERE originator_identity_id IS NOT NULL
        )
        UPDATE request_attempts
        SET attention_revision = (
          SELECT revision FROM ranked WHERE ranked.attempt_id = request_attempts.attempt_id
        )
        WHERE attempt_id IN (SELECT attempt_id FROM ranked);
        INSERT INTO request_attention_identities (identity_id, latest_revision, acknowledged_through)
          SELECT originator_identity_id, MAX(attention_revision), 0
          FROM request_attempts
          WHERE originator_identity_id IS NOT NULL
          GROUP BY originator_identity_id;
        CREATE INDEX request_attempts_attention
          ON request_attempts (originator_identity_id, attention_revision, request_id);
      `),
  },
];

const CREATE_MIGRATIONS = `
  CREATE TABLE IF NOT EXISTS _migrations (
    version INTEGER PRIMARY KEY,
    name TEXT NOT NULL,
    applied_at TEXT NOT NULL
  )
`;

function validateMigrationsTable(database: SqliteDatabase): void {
  const columns = database.pragma('table_info(_migrations)') as Array<{
    name: string;
    type: string;
    notnull: number;
    pk: number;
  }>;
  const expected = [
    { name: 'version', type: 'INTEGER', notnull: 0, pk: 1 },
    { name: 'name', type: 'TEXT', notnull: 1, pk: 0 },
    { name: 'applied_at', type: 'TEXT', notnull: 1, pk: 0 },
  ];
  if (
    columns.length !== expected.length ||
    columns.some((column, index) => {
      const wanted = expected[index];
      return (
        !wanted ||
        column.name !== wanted.name ||
        column.type.toUpperCase() !== wanted.type ||
        column.notnull !== wanted.notnull ||
        column.pk !== wanted.pk
      );
    })
  ) {
    throw incompatibleSchema('The _migrations table does not match the supported schema');
  }
}

function validateDefinitions(migrations: readonly MigrationDefinition[]): void {
  for (let index = 0; index < migrations.length; index += 1) {
    const migration = migrations[index];
    if (!migration || migration.version !== index + 1 || !migration.name || !migration.up) {
      throw incompatibleSchema('Migration definitions must be contiguous, forward-only, and named');
    }
  }
}

function validateHistory(
  history: Array<{ version: number; name: string }>,
  migrations: readonly MigrationDefinition[]
): number {
  let expected = 1;
  for (const row of history) {
    if (!Number.isSafeInteger(row.version) || row.version !== expected) {
      throw incompatibleSchema(`Migration history is not contiguous at version ${row.version}`);
    }
    const definition = migrations[row.version - 1];
    if (!definition) {
      throw incompatibleSchema(`Database requires unsupported migration version ${row.version}`);
    }
    if (definition.name !== row.name) {
      throw incompatibleSchema(
        `Migration ${row.version} has changed from ${row.name} to ${definition.name}`
      );
    }
    expected += 1;
  }
  return expected - 1;
}

export function applyMigrations(
  database: SqliteDatabase,
  migrations: readonly MigrationDefinition[] = CURRENT_MIGRATIONS
): number {
  validateDefinitions(migrations);

  try {
    database
      .transaction(() => {
        database.exec(CREATE_MIGRATIONS);
        validateMigrationsTable(database);
      })
      .immediate();

    const history = database
      .prepare('SELECT version, name FROM _migrations ORDER BY version')
      .all() as Array<{ version: number; name: string }>;
    const currentVersion = validateHistory(history, migrations);
    const insert = database.prepare(
      'INSERT INTO _migrations (version, name, applied_at) VALUES (?, ?, ?)'
    );

    for (let index = currentVersion; index < migrations.length; index += 1) {
      const migration = migrations[index];
      const applyOne = database.transaction(() => {
        try {
          // Recheck while holding the write transaction. Another process may
          // have applied this migration after the initial history read.
          const latest = database
            .prepare('SELECT version, name FROM _migrations ORDER BY version')
            .all() as Array<{ version: number; name: string }>;
          if (validateHistory(latest, migrations) >= migration.version) return;
          migration.up(database);
          insert.run(migration.version, migration.name, new Date().toISOString());
        } catch (error) {
          const classified = classifyStorageError(error, `Migration ${migration.version}`);
          throw new StorageError('migration', `Migration ${migration.version} failed`, {
            cause: classified,
            migrationVersion: migration.version,
            retryable: classified.retryable,
          });
        }
      });
      applyOne.immediate();
    }
    return migrations.length;
  } catch (error) {
    if (error instanceof StorageError) throw error;
    throw classifyStorageError(error, 'Database migration');
  }
}

File: rust/crates/tmt-adapters/src/storage/migrations.rs
use rusqlite::{Connection, TransactionBehavior, params};
use tmt_core::limits::MAX_JS_SAFE_INTEGER;

use super::errors::{StorageError, StorageErrorCode, classify, incompatible};

#[cfg(test)]
#[path = "identity_lifetime_tests.rs"]
mod identity_lifetime_tests;
#[cfg(test)]
mod receipt_tests;
#[cfg(test)]
mod test_support;

struct Migration {
    name: &'static str,
    sql: &'static str,
}

const MIGRATIONS: &[Migration] = &[
    Migration {
        name: "create durable identities and transient tmux bindings",
        sql: include_str!("schema/001.sql"),
    },
    Migration {
        name: "create optional identity role profiles",
        sql: include_str!("schema/002.sql"),
    },
    Migration {
        name: "create durable identity preambles",
        sql: include_str!("schema/003.sql"),
    },
    Migration {
        name: "create request attempts and preamble cadence counters",
        sql: include_str!("schema/004.sql"),
    },
    Migration {
        name: "create immutable request responses",
        sql: include_str!("schema/005.sql"),
    },
    Migration {
        name: "freeze exchange retention and response expiry horizons",
        sql: include_str!("schema/006_columns.sql"),
    },
    Migration {
        name: "retain request provenance and bounded original prompts",
        sql: include_str!("schema/007.sql"),
    },
    Migration {
        name: "add identity-scoped exchange attention revisions",
        sql: include_str!("schema/008.sql"),
    },
    Migration {
        name: "add identity lifetimes and reusable retired names",
        sql: include_str!("schema/009.sql"),
    },
];

pub(super) fn apply(connection: &mut Connection) -> Result<(), StorageError> {
    let transaction = connection
        .transaction_with_behavior(TransactionBehavior::Immediate)
        .map_err(|error| classify(error, "Initialize migration history"))?;
    transaction.execute_batch("CREATE TABLE IF NOT EXISTS _migrations (version INTEGER PRIMARY KEY, name TEXT NOT NULL, applied_at TEXT NOT NULL)")
        .map_err(|error| classify(error, "Initialize migration history"))?;
    validate_table(&transaction)?;
    transaction
        .commit()
        .map_err(|error| classify(error, "Commit migration history"))?;
    let current = validate_history(connection)?;
    for (index, migration) in MIGRATIONS.iter().enumerate().skip(current) {
        let version = index as u32 + 1;
        if version == 9 {
            apply_identity_lifetime(connection, migration)?;
        } else {
            apply_version(connection, version, migration)?;
        }
    }
    Ok(())
}

fn apply_identity_lifetime(
    connection: &mut Connection,
    migration: &Migration,
) -> Result<(), StorageError> {
    // SQLite requires foreign_keys to change outside a transaction.
    // The private opening connection cannot escape during this rebuild.
    let result = set_foreign_keys(connection, false)
        .map_err(|error| StorageError::migration(9, error))
        .and_then(|()| apply_version(connection, 9, migration));
    // apply_version has committed or dropped its transaction before
    // restoration. Always attempt it, preserving a primary failure.
    // Attempt restoration even if disabling succeeded but its verification failed.
    let restore =
        set_foreign_keys(connection, true).map_err(|error| StorageError::migration(9, error));
    result.and(restore)
}

fn apply_version(
    connection: &mut Connection,
    version: u32,
    migration: &Migration,
) -> Result<(), StorageError> {
    let transaction = connection
        .transaction_with_behavior(TransactionBehavior::Immediate)
        .map_err(|error| classify(error, "Acquire migration lock"))?;
    let apply_one = (|| {
        // Another opener may have migrated while this one waited for the lock.
        if validate_history(&transaction)? >= version as usize {
            return Ok(());
        }
        if version == 9 {
            validate_identity_source(&transaction)?;
            check_foreign_keys(&transaction)?;
        }
        transaction
            .execute_batch(migration.sql)
            .map_err(|error| classify(error, "Apply migration SQL"))?;
        if version == 6 {
            backfill_retention(&transaction)?;
        }
        transaction.execute(
            "INSERT INTO _migrations (version, name, applied_at) VALUES (?, ?, strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))",
            params![version, migration.name],
        ).map_err(|error| classify(error, "Record migration"))?;
        if version == 9 {
            check_foreign_keys(&transaction)?;
        }
        Ok(())
    })();
    apply_one.map_err(|error| StorageError::migration(version, error))?;
    transaction
        .commit()
        .map_err(|error| classify(error, "Commit migration"))
}

fn set_foreign_keys(connection: &Connection, enabled: bool) -> Result<(), StorageError> {
    connection
        .pragma_update(None, "foreign_keys", enabled)
        .map_err(|error| classify(error, "Configure migration foreign keys"))?;
    let actual: bool = connection
        .pragma_query_value(None, "foreign_keys", |row| row.get(0))
        .map_err(|error| classify(error, "Verify migration foreign keys"))?;
    if actual != enabled {
        return Err(incompatible(
            "Migration foreign-key policy did not take effect",
        ));
    }
    Ok(())
}

fn check_foreign_keys(connection: &Connection) -> Result<(), StorageError> {
    let mut statement = connection
        .prepare("PRAGMA foreign_key_check")
        .map_err(|error| classify(error, "Inspect migration foreign keys"))?;
    let mut rows = statement
        .query([])
        .map_err(|error| classify(error, "Inspect migration foreign keys"))?;
    if rows
        .next()
        .map_err(|error| classify(error, "Read migration foreign keys"))?
        .is_some()
    {
        return Err(incompatible(
            "Identity migration requires consistent foreign keys",
        ));
    }
    Ok(())
}

fn validate_identity_source(connection: &Connection) -> Result<(), StorageError> {
    // Compare only the frozen first CREATE TABLE statement, not arbitrary SQL.
    // Whitespace differs between historical TS and Rust migration formatting.
    // Refuse custom columns/constraints rather than discarding their data/rules.
    let original = MIGRATIONS[0].sql.split(';').next().unwrap_or_default();
    let actual: String = connection
        .query_row(
            "SELECT sql FROM sqlite_schema WHERE type = 'table' AND name = 'identities'",
            [],
            |row| row.get(0),
        )
        .map_err(|error| classify(error, "Inspect identity source schema"))?;
    if !actual.split_whitespace().eq(original.split_whitespace()) {
        return Err(incompatible(
            "Identity migration requires the historical table definition",
        ));
    }
    // Only the two implicit identity indexes are part of schema 8. Never drop
    // an unrecognized user index or trigger as a side effect of table replacement.
    let custom: bool = connection.query_row(
        "SELECT EXISTS(SELECT 1 FROM sqlite_schema WHERE tbl_name = 'identities' AND type IN ('index', 'trigger') AND sql IS NOT NULL)",
        [], |row| row.get(0),
    ).map_err(|error| classify(error, "Inspect identity schema extensions"))?;
    if custom {
        return Err(incompatible(
            "Identity migration cannot replace custom indexes or triggers",
        ));
    }
    Ok(())
}

fn validate_table(connection: &Connection) -> Result<(), StorageError> {
    let mut query = connection
        .prepare("PRAGMA table_info(_migrations)")
        .map_err(|error| classify(error, "Inspect migration schema"))?;
    let columns = query
        .query_map([], |row| {
            Ok((
                row.get::<_, String>(1)?,
                row.get::<_, String>(2)?,
                row.get::<_, i64>(3)?,
                row.get::<_, i64>(5)?,
            ))
        })
        .map_err(|error| classify(error, "Inspect migration schema"))?
        .collect::<rusqlite::Result<Vec<_>>>()
        .map_err(|error| classify(error, "Read migration schema"))?;
    let expected = [
        ("version", "INTEGER", 0, 1),
        ("name", "TEXT", 1, 0),
        ("applied_at", "TEXT", 1, 0),
    ];
    if columns.len() != expected.len()
        || columns.iter().zip(expected).any(|(actual, expected)| {
            actual.0 != expected.0
                || actual.1.to_ascii_uppercase() != expected.1
                || actual.2 != expected.2
                || actual.3 != expected.3
        })
    {
        return Err(incompatible(
            "The _migrations table does not match the supported schema",
        ));
    }
    Ok(())
}

fn validate_history(connection: &Connection) -> Result<usize, StorageError> {
    let mut statement = connection
        .prepare("SELECT version, name FROM _migrations ORDER BY version")
        .map_err(|error| classify(error, "Read migration history"))?;
    let rows = statement
        .query_map([], |row| {
            Ok((row.get::<_, i64>(0)?, row.get::<_, String>(1)?))
        })
        .map_err(|error| classify(error, "Read migration history"))?;
    let mut count = 0;
    for row in rows {
        let (version, name) = row.map_err(|error| {
            incompatible("Migration history contains invalid values").caused_by(error)
        })?;
        if version != count as i64 + 1 {
            return Err(incompatible(format!(
                "Migration history is not contiguous at version {version}"
            )));
        }
        let Some(definition) = MIGRATIONS.get(count) else {
            return Err(incompatible(format!(
                "Database requires unsupported migration version {version}"
            )));
        };
        if definition.name != name {
            return Err(incompatible(format!(
                "Migration {version} has changed from {name} to {}",
                definition.name
            )));
        }
        count += 1;
    }
    Ok(count)
}

// Historical migration policy is frozen independently of today's settings.
const LEGACY_RETENTION_MS: i64 = 7 * 86_400_000;
const SETTLEMENT_FLOOR_MS: i64 = 86_400_000;

fn saturating_deadline(value: i64, delta: i64) -> Result<i64, StorageError> {
    let maximum = MAX_JS_SAFE_INTEGER as i64;
    if !(1..=maximum).contains(&value) || !(0..=maximum).contains(&delta) {
        return Err(StorageError::new(
            StorageErrorCode::Unknown,
            "Historical timestamp is outside the supported range",
        ));
    }
    Ok(value.saturating_add(delta).min(maximum))
}

fn backfill_retention(connection: &Connection) -> Result<(), StorageError> {
    let mut first = connection.prepare("SELECT attempt_id, prepared_at_ms, expires_at_ms, settled_at_ms FROM request_attempts ORDER BY attempt_id LIMIT 100")
        .map_err(|error| classify(error, "Prepare retention migration"))?;
    let mut next = connection.prepare("SELECT attempt_id, prepared_at_ms, expires_at_ms, settled_at_ms FROM request_attempts WHERE attempt_id > ? ORDER BY attempt_id LIMIT 100")
        .map_err(|error| classify(error, "Prepare retention migration"))?;
    let mut update = connection
        .prepare("UPDATE request_attempts SET retention_expires_at_ms = ? WHERE attempt_id = ?")
        .map_err(|error| classify(error, "Prepare retention migration"))?;
    let mut previous: Option<String> = None;
    loop {
        let read = |row: &rusqlite::Row<'_>| {
            Ok((
                row.get::<_, String>(0)?,
                row.get::<_, i64>(1)?,
                row.get::<_, i64>(2)?,
                row.get::<_, Option<i64>>(3)?,
            ))
        };
        let rows = match &previous {
            Some(previous) => next.query_map([previous], read),
            None => first.query_map([], read),
        }
        .map_err(|error| classify(error, "Read historical attempts"))?
        .collect::<rusqlite::Result<Vec<_>>>()
        .map_err(|error| classify(error, "Decode historical attempts"))?;
        if rows.is_empty() {
            break;
        }
        for (id, prepared, expires, settled) in rows {
            // The historical reply-acceptance window is also seven days.
            let horizon = saturating_deadline(prepared, LEGACY_RETENTION_MS)?
                .max(saturating_deadline(expires, SETTLEMENT_FLOOR_MS)?)
                .max(
                    settled
                        .map(|time| saturating_deadline(time, SETTLEMENT_FLOOR_MS))
                        .transpose()?
                        .unwrap_or(0),
                );
            update
                .execute(params![horizon, id])
                .map_err(|error| classify(error, "Backfill attempt horizon"))?;
            previous = Some(id);
        }
    }

    let mut first = connection.prepare("SELECT request_id, submitted_at_ms FROM request_responses ORDER BY request_id LIMIT 100")
        .map_err(|error| classify(error, "Prepare response migration"))?;
    let mut next = connection.prepare("SELECT request_id, submitted_at_ms FROM request_responses WHERE request_id > ? ORDER BY request_id LIMIT 100")
        .map_err(|error| classify(error, "Prepare response migration"))?;
    let mut update = connection
        .prepare("UPDATE request_responses SET response_expires_at_ms = ? WHERE request_id = ?")
        .map_err(|error| classify(error, "Prepare response migration"))?;
    let mut previous: Option<String> = None;
    loop {
        let read = |row: &rusqlite::Row<'_>| Ok((row.get::<_, String>(0)?, row.get::<_, i64>(1)?));
        let rows = match &previous {
            Some(previous) => next.query_map([previous], read),
            None => first.query_map([], read),
        }
        .map_err(|error| classify(error, "Read historical responses"))?
        .collect::<rusqlite::Result<Vec<_>>>()
        .map_err(|error| classify(error, "Decode historical responses"))?;
        if rows.is_empty() {
            break;
        }
        for (id, submitted) in rows {
            update
                .execute(params![
                    saturating_deadline(submitted, LEGACY_RETENTION_MS)?,
                    id
                ])
                .map_err(|error| classify(error, "Backfill response horizon"))?;
            previous = Some(id);
        }
    }
    connection.execute_batch("UPDATE request_attempts SET retention_expires_at_ms = MAX(retention_expires_at_ms, COALESCE((SELECT response_expires_at_ms FROM request_responses WHERE request_responses.request_id = request_attempts.request_id AND request_responses.attempt_id = request_attempts.attempt_id), retention_expires_at_ms))")
        .map_err(|error| classify(error, "Extend matching response horizons"))?;
    connection
        .execute_batch(include_str!("schema/006_indexes.sql"))
        .map_err(|error| classify(error, "Index retention horizons"))?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn historical_deadlines_saturate_but_do_not_accept_invalid_anchors() {
        assert_eq!(
            saturating_deadline(1, LEGACY_RETENTION_MS).unwrap(),
            604_800_001
        );
        assert_eq!(
            saturating_deadline(MAX_JS_SAFE_INTEGER as i64 - 1, LEGACY_RETENTION_MS).unwrap(),
            MAX_JS_SAFE_INTEGER as i64
        );
        for value in [i64::MIN, -1, 0, MAX_JS_SAFE_INTEGER as i64 + 1, i64::MAX] {
            assert_eq!(
                saturating_deadline(value, 1).unwrap_err().code,
                StorageErrorCode::Unknown
            );
        }
    }
}
